virtualizarr/readers/safetensors.py

```diff
@@ -0,0 +1,641 @@
+        ManifestArrays for each tensor. Each tensor is treated as a single chunk for
+        efficient memory-mapped access.
+
+        Parameters
+        ----------
+        filepath : str
+            Path to the SafeTensors file. Can be a local path or a URL.
+        drop_variables : list
+            List of tensor names to exclude from the dataset.
+        store: ObjectStore
+            Object store used for reading the file. Should match the filepath type.
+        dimension_names : Dict[str, list[str]], optional
+            Custom dimension names for specific tensors. The keys should be tensor names,
+            and the values should be lists of dimension names matching the tensor's shape.
+            If not provided, default names are generated as "{tensor_name}_dim_{i}".
+
+        Returns
+        -------
+        ManifestGroup
+            A group containing ManifestArrays for each tensor, with appropriate
+            metadata and attributes.
+
+        Notes
+        -----
+        - Each tensor is represented as a single chunk for direct memory access
+        - The __metadata__ field from the SafeTensors header is preserved as attributes
+        - Tensor metadata includes:
+          - Original SafeTensors dtype (e.g., "F32", "I64", "BF16")
+          - Storage information indicating contiguous layout
+          - Any additional tensor-specific metadata from the header
+
+        Examples
+        --------
+        >>> store = default_object_store("model.safetensors")
+        >>> dimension_names = {
+        ...     "weight": ["input_channels", "output_channels"],
+        ...     "bias": ["features"]
+        ... }
+        >>> manifest_group = _create_manifest_group(
+        ...     "model.safetensors",
+        ...     drop_variables=[],
+        ...     store=store,
+        ...     dimension_names=dimension_names
+        ... )
+        """
+        header, header_size = SafeTensorsVirtualBackend._parse_safetensors_header(
+            filepath, store
+        )
+
+        manifest_dict = {}
+
+        attrs = {}
+        if "__metadata__" in header:
+            metadata_content = header["__metadata__"]
+            if isinstance(metadata_content, dict):
+                for key, value in metadata_content.items():
+                    # safetensors spec only allows text-to-text map in __metadata__
+                    if not isinstance(value, str):
+                        value = json.dumps(value)
+                    attrs[key] = value
+            else:
+                attrs["__metadata__"] = (
+                    metadata_content
+                    if isinstance(metadata_content, str)
+                    else json.dumps(metadata_content)
+                )
+
+        data_start = 8 + header_size
+
+        for tensor_name, tensor_info in header.items():
+            if tensor_name == "__metadata__" or tensor_name in drop_variables:
+                continue
```

Comment:
It would be a little neater to abstract the logic inside this loop into a function.